src/thunderbots/software/ai/passing/evaluation.cpp

```diff
@@ -2,14 +2,295 @@
+}
+
+double AI::Passing::calculateInterceptRisk(const Team& enemy_team, const Pass& pass)
+{
+    // Return the highest risk for all the enemy robots, if there are any
+    auto enemy_robots = enemy_team.getAllRobots();
+    if (enemy_robots.empty())
+    {
+        return 0;
+    }
+    std::vector<double> enemy_intercept_risks;
+    std::transform(enemy_robots.begin(), enemy_robots.end(),
+                   std::back_inserter(enemy_intercept_risks),
+                   [&](Robot robot) { return calculateInterceptRisk(robot, pass); });
+    return *std::max_element(enemy_intercept_risks.begin(), enemy_intercept_risks.end());
+}
+
+double AI::Passing::calculateInterceptRisk(Robot enemy_robot, const Pass& pass)
+{
+    // We estimate the intercept by the risk that the robot will get to the closest
+    // point on the pass before the ball, and by the risk that the robot will get to
+    // the reception point before the ball. We take the greater of these two risks.
+    // We assume that the enemy continues moving at it's current velocity until the
+    // pass starts
+
+    // TODO: Some of the variable names here are real long and a bit ugly
+
+    // We force any negative duration to 0 here
+    Duration time_until_pass = pass.startTime() - enemy_robot.lastUpdateTimestamp();
+
+    // Estimate where the enemy will be when we start the pass
+    enemy_robot.updateStateToPredictedState(time_until_pass);
+
+    // If the enemy cannot intercept the pass at BOTH the closest point on the pass and
+    // the the receiver point for the pass, then it is guaranteed that it will not be
+    // able to intercept the pass anywhere.
+
+    // Figure out how long the enemy robot and ball will take to reach the closest
+    // point on the pass to the enemy's current position
+    Point closest_point_on_pass_to_robot = closestPointOnSeg(
+        enemy_robot.position(), pass.passerPoint(), pass.receiverPoint());
+    Duration enemy_robot_time_to_closest_pass_point =
+        getTimeToPositionForRobot(enemy_robot, closest_point_on_pass_to_robot);
+    Duration ball_time_to_closest_pass_point = Duration::fromSeconds(
+        (closest_point_on_pass_to_robot - pass.passerPoint()).len() / pass.speed());
+
+    // Figure out how long the enemy robot and ball will take to reach the receive point
+    // for the pass.
+    Duration enemy_robot_time_to_pass_receive_position =
+        getTimeToPositionForRobot(enemy_robot, pass.receiverPoint());
+    Duration ball_time_to_pass_receive_position = pass.estimatePassDuration();
+
+    double robot_ball_time_diff_at_closest_pass_point =
+        (enemy_robot_time_to_closest_pass_point - ball_time_to_closest_pass_point)
+            .getSeconds();
+    double robot_ball_time_diff_at_pass_receive_point =
+        (enemy_robot_time_to_pass_receive_position - ball_time_to_pass_receive_position)
+            .getSeconds();
+
+    // We take a smooth "max" of these two values using a log-sum-exp function
```

Comment:
probably a good idea to leave a comment with the link to the log-sum-exp resource you used, especially since this is the "computationally stable" version.

Reply:
added